File: Algorithms/030.SubstringWithConcatenationOfAllWords.cpp
#include "leetcode.h"
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> result;
        if (words.size() == 0)
            return result;
        int wordLength = words[0].size();
        int subLength = wordLength * words.size();
        int length = s.size();
        unordered_map<string, int>  wordCount, current;
        for (int i=0; i < words.size(); ++i)
            wordCount[words[i]] ++;
        for (int i=0; i <= length-subLength; i++) {
            current.clear();
            int j;
            for (j=0; j < words.size(); j++) {
                string word = s.substr(i + j*wordLength, wordLength);
                if (wordCount.find(word) == wordCount.end())
                    break;
                current[word] ++;
                if (current[word] > wordCount[word])
                    break;
            }
            if (j == words.size())
                result.push_back(i);
        }
        return result;
    }
};
int main() {
    Solution s;
    string S = "barfoothefoobarman";
    vector<string> L;
    L.push_back("foo");
    L.push_back("bar");
    vector<int> result = s.findSubstring(S, L);
    for (int i=0; i < result.size(); i++)
        cout << result[i] << " ";
    cout << endl;
    return 0;
}

File: Algorithms/py/030.SubstringWithConcatenationOfAllWords.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        result = []
        wordLength = len(words[0])
        wordCount = collections.defaultdict(int)
        current = collections.defaultdict(int)
        for word in words:
            wordCount[word] += 1
        for i in range(len(s)-wordLength * len(words)+1):
            current.clear();
            for j in range(len(words)):
                word = s[i + j*wordLength: i + (j+1)*wordLength]
                if word not in wordCount:
                    break
                current[word] += 1
                if current[word] > wordCount[word]:
                    break
            else:
                result.append(i)
        return result